treescript/src/treescript/l10n.py

```diff
@@ -77,26 +81,82 @@ def build_platform_dict(bump_config, repo_path):
+    last_push_id = revision_info["lastpushid"]
+    revision = revision_info["pushes"][str(last_push_id)]["changesets"][0]
+    log.info(f"locale {locale} revision {revision}")
+    return (locale, revision)
+
+
 # build_revision_dict {{{1
-def build_revision_dict(l10n_bump_info, revision_info, repo_path):
+async def build_revision_dict(bump_config, repo_path, old_contents, dashboard_revision_info=None):
     """Add l10n revision information to the ``platform_dict``.
 
-    The l10n dashboard contains locale to revision information for each
-    locale. If we have a ``revision_url``, that is the templatized dashboard
-    url we should query for locale to revision information.
+    If we have an ``l10n_repo_url``, use that as a template for the locale
+    repo url. If ``pin`` is set in the ``old_contents`` for that locale, save
+    the previous revision and pin. Otherwise, find the latest revision in the
+    locale repo, and use that.
+
+    The ``l10n_repo_url`` will look something like
+    https://hg.mozilla.org/l10n-central/%(locale)s/json-pushes?version=2&tipsonly=1
+
+    Deprecated: the l10n dashboard contains locale to revision information for
+    each locale. If we have a ``dashboard_revision_url``, that is the
+    templatized dashboard url we should query for locale to revision
+    information.
 
     Otherwise, add a ``default`` revision to each locale in the
     ``platform_dict``.
 
+    Args:
+        bump_config (dict): one of the dictionaries from the
+            payload ``l10n_bump_info``
+        repo_path (str): the path to the repo on disk
+        old_contents (dict): the old contents of the l10n changesets, if any.
+        dashboard_revision_info (string, optional): the contents of the
+            l10n dashboard. Defaults to None.
+
     Returns:
         dict: locale to dictionary of platforms and revision
 
     """
     log.info("Building revision dict...")
-    platform_dict = build_platform_dict(l10n_bump_info, repo_path)
+    platform_dict = build_platform_dict(bump_config, repo_path)
     revision_dict = {}
-    if revision_info:
-        for line in revision_info.splitlines():
+    if bump_config.get("l10n_repo_url"):
+        semaphore = asyncio.Semaphore(5)
+        tasks = []
+        for locale, value in platform_dict.items():
+            if old_contents.get(locale, {}).get("pin"):
+                value["revision"] = old_contents[locale]["revision"]
+                value["pin"] = old_contents[locale]["pin"]
+            else:
+                tasks.append(asyncio.create_task(semaphore_wrapper(semaphore, get_latest_revision(locale, bump_config["l10n_repo_url"]))))
+                value["pin"] = False
+            revision_dict[locale] = value
+        await asyncio.gather(*tasks)
+        for task in tasks:
+            (locale, revision) = task.result()
+            revision_dict[locale]["revision"] = revision
+    elif dashboard_revision_info:
```

Comment:
can we kill this branch of the code?

Reply:
I think beta and esr68 will use this branch of code til a) central-> beta merge day, and b) we turn off the bumper in esr68. Once those happen, then yes, I'd love to clean this up.